synapse/handlers/room_list.py

```diff
@@ -156,62 +156,99 @@ async def _get_public_room_list(
         # plus the direction (forwards or backwards). Next batch tokens always
         # go forwards, prev batch tokens always go backwards.
 
+        forwards = True
+        last_joined_members = None
+        last_room_id = None
+        last_module_index = None
         if since_token:
             batch_token = RoomListNextBatch.from_token(since_token)
             forwards = batch_token.direction_is_forward
-        else:
-            batch_token = None
-            forwards = True
+            last_joined_members = batch_token.last_joined_members
+            last_room_id = batch_token.last_room_id
+            last_module_index = batch_token.last_module_index
 
         # we request one more than wanted to see if there are more pages to come
         probing_limit = limit + 1 if limit is not None else None
 
-        results = await self.store.get_largest_public_rooms(
-            network_tuple,
-            search_filter,
-            probing_limit,
-            bounds=(batch_token.last_joined_members, batch_token.last_room_id)
-            if batch_token
-            else None,
-            forwards=forwards,
-            ignore_non_federatable=bool(from_remote_server_name),
-        )
+        results = []
+
+        print(f"last_module_index {last_module_index}")
+        print(f"last_room_id {last_room_id}")
+
+        def insert_into_result(new_room: PublicRoom, module_index: Optional[int]):
+            # print(f"insert {new_room.room_id} {module_index}")
+            if new_room.num_joined_members == last_joined_members:
+                if last_module_index is not None and last_room_id is not None:
+                    if module_index is not None and module_index > last_module_index:
+                        return
+            inserted = False
+            for i, r in enumerate(results):
+                r = results[i]
+                if (
+                    forwards and new_room.num_joined_members >= r.num_joined_members
+                ) or (
+                    not forwards and new_room.num_joined_members <= r.num_joined_members
+                ):
+                    results.insert(i, new_room)
+                    inserted = True
+                    return
+            if not inserted:
+                if forwards:
+                    results.append(new_room)
+                else:
+                    results.insert(0, new_room)
```

Comment:
Would it be easier to first build an unsorted list of `(Optional[module_index], PublicRoom)` tuples, then sort it and slice off `limit` entries? This insertion sort-type logic looks quite tricky to get right.

Reply:
`limit` is already dealt with global slicing a bit later.

We need to keep the order returned by the modules when sorting so it is repeatable and pagination works. I should be able to do that by adding an a per module index to the tuple however.
I'll try that, because I agree with you that this is confusing code.

Thanks for the early look.